crates/red_knot_python_semantic/src/visibility_constraints.rs

```diff
@@ -0,0 +1,180 @@
+use ruff_index::IndexVec;
+
+use crate::semantic_index::{
+    ast_ids::HasScopedExpressionId,
+    constraint::{Constraint, ConstraintNode, PatternConstraintKind},
+    AllConstraints,
+};
+use crate::semantic_index::{ScopedConstraintId, ScopedVisibilityConstraintId};
+use crate::types::{infer_expression_types, Truthiness};
+use crate::Db;
+
+const MAX_RECURSION_DEPTH: usize = 10;
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub(crate) enum VisibilityConstraint {
+    AlwaysTrue,
+    VisibleIf(ScopedConstraintId),
+    VisibleIfNot(ScopedVisibilityConstraintId),
+    KleeneAnd(ScopedVisibilityConstraintId, ScopedVisibilityConstraintId),
+    KleeneOr(ScopedVisibilityConstraintId, ScopedVisibilityConstraintId),
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub(crate) struct VisibilityConstraints {
+    constraints: IndexVec<ScopedVisibilityConstraintId, VisibilityConstraint>,
```

Comment:
I started the experiment here, but gave up after ~1.5 hours with a lot of lifetime errors still to be resolved: https://github.com/astral-sh/ruff/pull/15079

Reply:
Oof. Yeah I don't think we should spend more time on this now, since I think it would may become irrelevant if we switch to a BDD representation of the constraint tree.